## **Task-2 :    CREDIT CARD FRAUD DETECTION**

## **Author: Yuvraj Verma**
### Batch October
### Domain: Data Science
### Aim: To Build a model to predict Credit Card Fraud Transactions

## Importing Deendencies

In [ ]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

## Loading the dataset to a Pandas Dataframe

In [ ]:
credit_card_data = pd.read_csv("/content/drive/MyDrive/CodeSoft/creditcard.csv")

## First five rows of the Dataset

In [ ]:
credit_card_data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [ ]:
credit_card_data.tail()

Time         V1         V2        V3        V4        V5  \
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[5 rows x 31 columns]

## Information Of Dataset

In [ ]:
credit_card_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

## Checking the number of missing values in each column

In [ ]:
credit_card_data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

## Distribution of legit & Fraudulent transaction

In [ ]:
credit_card_data['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

### This dataset is highly unbalance.

### **0-->Normal Transaction**
### **1-->Fraudulent Transaction**

### Searating the data for analysis

In [ ]:
legit = credit_card_data[credit_card_data.Class == 0]
fraud = credit_card_data[credit_card_data.Class ==1]

In [ ]:
print(legit.shape)
print(fraud.shape)

(284315, 31)
(492, 31)


### **Statistical measure for analysis**

In [ ]:
legit.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [ ]:
fraud.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

### **Campare the values for both transaction**

In [ ]:
credit_card_data.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

## **Under-Sampling**

### Build a sample dataset contaning similar distribution of naormal transaction and fraudulant transaction

### Numbers of Fraudulant transaction -->492

In [ ]:
legit_sample = legit.sample(n=492)

### Concatenating two Dataframes

In [ ]:
new_dataset = pd.concat([legit_sample, fraud], axis=0)

In [ ]:
new_dataset.head()

Time         V1         V2        V3        V4        V5  \
243521  151964.0   1.855972  -1.664855 -0.515613  0.060069  0.531164   
74185    55460.0 -11.746656 -10.410737 -1.273259  5.588220 -2.582046   
224230  143703.0  -0.324718   1.352248 -0.782295 -0.327054  1.464372   
187190  127429.0  -0.809139  -0.284447  0.043106 -0.203763 -2.284870   
230842  146505.0   2.179893  -1.030659 -2.653366 -1.894402  1.879562   

              V6        V7        V8        V9  ...       V21       V22  \
243521  5.238788 -2.463713  1.470498  1.204216  ... -0.221603  0.209841   
74185   5.319379  6.079441 -4.303163  4.475720  ... -4.061621  1.044273   
224230 -0.848050  1.212435 -0.178671 -0.106258  ...  0.009352  0.438133   
187190  2.235805  2.748954 -0.656769  0.905119  ... -0.178935  0.524745   
230842  3.251897 -0.941101  0.744315 -0.492534  ... -0.029570 -0.150150   

             V23       V24       V25       V26       V27       V28   Amount  \
243521  0.207020  0.686831 -0.217465 -0.455934  0.172520 -0.012130    50.00   
74185  -4.415716 -0.929446  0.180019  0.436373 -3.420042  6.649828  1580.56   
224230 -0.449073 -0.832637  0.085786  0.721037  0.381717  0.271183    12.31   
187190 -0.046763  0.127148 -1.403498  0.047732  0.210306 -0.091242   652.18   
230842  0.261500  0.706477 -0.039307 -0.307677 -0.013611 -0.065775    19.50   

        Class  
243521      0  
74185       0  
224230      0  
187190      0  
230842      0  

[5 rows x 31 columns]

In [ ]:
new_dataset.tail()

Time        V1        V2        V3        V4        V5        V6  \
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[5 rows x 31 columns]

In [ ]:
new_dataset['Class'].value_counts()

0    492
1    492
Name: Class, dtype: int64

In [ ]:
new_dataset.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      95159.394309  0.025424 -0.061947  0.043867  0.042646  0.018785   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.020225  0.040401  0.048515  0.067438  ... -0.062244 -0.061386   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0      0.007987  0.004002 -0.015504  0.030351  0.007189 -0.014221  0.047847   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       83.748476  
1      122.211321  

[2 rows x 30 columns]

### **Splitting the data into Features & targets**

In [ ]:
X = new_dataset.drop(columns='Class', axis = 1)
Y = new_dataset['Class']

In [ ]:
print(X)

            Time         V1         V2        V3        V4        V5  \
243521  151964.0   1.855972  -1.664855 -0.515613  0.060069  0.531164   
74185    55460.0 -11.746656 -10.410737 -1.273259  5.588220 -2.582046   
224230  143703.0  -0.324718   1.352248 -0.782295 -0.327054  1.464372   
187190  127429.0  -0.809139  -0.284447  0.043106 -0.203763 -2.284870   
230842  146505.0   2.179893  -1.030659 -2.653366 -1.894402  1.879562   
...          ...        ...        ...       ...       ...       ...   
279863  169142.0  -1.927883   1.125653 -4.518331  1.749293 -1.566487   
280143  169347.0   1.378559   1.289381 -5.004247  1.411850  0.442581   
280149  169351.0  -0.676143   1.126366 -2.213700  0.468308 -1.120541   
281144  169966.0  -3.113832   0.585864 -5.399730  1.817092 -0.840618   
281674  170348.0   1.991976   0.158476 -2.583441  0.408670  1.151147   

              V6        V7        V8        V9  ...        V20       V21  \
243521  5.238788 -2.463713  1.470498  1.204216  ...  -0.522

In [ ]:
print(Y)

243521    0
74185     0
224230    0
187190    0
230842    0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64


### **Split the data into training data and testing data**

In [ ]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, stratify = Y, random_state=2)

In [ ]:
print(X.shape, X_train.shape, X_test.shape)

(984, 30) (787, 30) (197, 30)


# **Model Training**

### Logistic Regression

In [ ]:
model = LogisticRegression()

### Trainig the Logistic Regression Model with Training Data

In [ ]:
model.fit(X_train, Y_train)

LogisticRegression()

## **Model Evaluation**

### Accuracy Score

In [ ]:
#Accuracy on training data

X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)

In [ ]:
print('Accuracy on Training data : ', training_data_accuracy)

Accuracy on Training data :  0.9428208386277002


In [ ]:
#Accuracy on test data
X_test_prediction = model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)

In [ ]:
print('Accuracy score on Test Data : ',test_data_accuracy)

Accuracy score on Test Data :  0.949238578680203


#### **Therfor the accuracy on Training Data is :  0.9428208386277002**
#### **The accuracy on Ttesting Data is: 0.949238578680203**

# **Thankyou**